### DISCUSSION TITLE
Debugging queries by exploring types

### DISCUSSION THREAD

**Discussion by mtardy** [github/codeql#9895] [Q&A]
I'm quite new to CodeQL but the tool is really impressive for tracing data flow in programs.
I did the whole tutorial about the abstract situation with the villagers and everything but when switching to real Golang use cases, it start to become difficult.

Is there any way of seeing/debugging/printing all the data associated with some types?

For example, I wanted to see all the things related to `InterfaceType` in my program, or more generally, interfaces, because I wanted to detect the implementation of some interface and it's quite frustrating to have to `select` each possible method of the different types to see if there are some related data that exists to filter on. It seems it is really difficult to browse the types to understand how to filter them.

Do you have any piece of advice related to this? Dumping all available information on a type to see it?

**Comment by alexet** [comment]
I don't think we have anything that does precisely what you want. For the AST you can use the ast viewer to see the representation of things in the AST but that doesn't apply to types.

Within vscode you can highlight the method name and then use `CodeQL: Quick Evaluation` to evaluate it without having to write a full query whcih may save you time.This also applies to full expressions and formulas.

To just easily see which methods are available we have reference documentation here: `https://codeql.github.com/codeql-standard-libraries/go/semmle/go/Types.qll/type.Types$InterfaceType.html`

**Reply by mtardy** [reply]
Yes, I actually used the AST and the quick eval and it's great. But it's pretty hard to have a whole visualization of types without trying. For the methods, for sure, the documentation is nice.

Anyway, thanks I wanted to make sure I was not missing something! :) I will fight with those types to find what I want and see how they are built.
